src/bartiq/analysis/rewriters/expression_rewriter.py

```diff
@@ -134,45 +147,62 @@ def individual_terms(self) -> Iterable[T]:
         """Return the expression as an iterable of individual terms."""
 
     @abstractmethod
-    def focus(self, symbols: str | Iterable[str]) -> TExpr[T]:
+    def focus(self, symbols: str | Iterable[str]) -> T:
         """Return an expression containing terms that involve specific symbols."""
 
     @abstractmethod
-    def _expand(self) -> TExpr[T]:
-        pass
+    def _expand(self) -> T: ...
 
     def expand(self) -> Self:
         """Expand all brackets in the expression."""
-        return replace(self, expression=self._expand(), _previous=(Instruction.Expand, self))
+        return replace(self, expression=self._expand(), _previous=(Expand(), self))
 
     @abstractmethod
-    def _simplify(self) -> TExpr[T]:
-        pass
+    def _simplify(self) -> T: ...
 
     def simplify(self) -> Self:
         """Run the backend `simplify' functionality, if it exists."""
-        return replace(self, expression=self._simplify(), _previous=(Instruction.Simplify, self))
+        return replace(self, expression=self._simplify(), _previous=(Simplify(), self))
 
     @abstractmethod
-    def _assume(self, assumption: str | Assumption) -> TExpr[T]:
-        pass
+    def _assume(self, assumption: Assumption) -> T: ...
 
     def assume(self, assumption: str | Assumption) -> Self:
         """Add an assumption for a symbol."""
         assumption = Assumption.from_string(assumption) if isinstance(assumption, str) else assumption
         return replace(
             self,
             expression=self._assume(assumption=assumption),
-            assumptions=self.assumptions + (assumption,),
-            _previous=(str(assumption), self),
+            _previous=(assumption, self),
         )
 
     def reapply_all_assumptions(self) -> Self:
         """Reapply all previously applied assumptions."""
         current = self
         for assumption in self.assumptions:
             current = current.assume(assumption=assumption)
-        return replace(self, expression=current.expression, _previous=(Instruction.ReapplyAllAssumptions, self))
+        return replace(self, expression=current.expression, _previous=(ReapplyAllAssumptions(), self))
+
+    @abstractmethod
+    def _substitute(self, substitution: Substitution) -> T: ...
+
+    def substitute(self, symbol_or_expr: str, replace_with: str) -> Self:
+        """Substitute a symbol or subexpression for another symbol or subexpression.
+        By default performs a one-to-one mapping, unless wildcard symbols are implemented.
```

Comment:
[minor] "unless wildcard symbols are implemented. -> "unless there are wildcards present in `replace_with`" ? I'm not entirely sure what you meant here.

Also, it looks to me that this docstring (and some other as well) are missing `Args` section :)